MetaMask/metamask-extension
# e2e: add Snap Accounts ethSign test, convert to TypeScript, optimize

## **Description**

Continuing to follow this E2E development plan https://www.notion.so/Snaps-Keyring-V1-290f59e37414470d97030e5dc22ab36d

### Important Features
- The Eth Sign tests have been added
- The whole `test/e2e/accounts` folder is now in TypeScript.  As far as I can tell, this creates Extension's first TypeScript E2E tests.  (This was not an easy task, because there was no prior work to follow, and I hit a bunch of type and import hurdles along the way.)  To support this:
  - Updated `mocha` to the newest `10.2.0`
  - Added `@types/mocha` and `@types/selenium-webdriver`
  - Required some changes in `test/e2e/webdriver/driver.js`, most notably how the `PAGES` object works
  - JSDoc in some JS files to get better type hints when there's no TypeScript
- `test-snap-accounts.spec.js` was previously the longest-runtime E2E testfile, often taking 18-25 minutes all by itself, and being the bottleneck.  I split and optimized the tests, so now the longest-runtime testfile is `snap-account-signatures.spec.ts` at 4 minutes.
- Stabilized a bunch of E2E tests using these techniques:
  - No more `findElement` then `assert(getText())` pattern
  - Safer window waiting and switching
  - Using `clickElementSafe` for scroll buttons
- Made `test-e2e-chrome-multichain` required for `all-tests-pass` again.  I did not actually fix the E2E tests that were failing, I just added this code to another spot in `snap-accounts-transfers.spec.ts`.  I'm not sure that actually fixing it is possible right now.
```
// TODO: Update Test when Multichain Send Flow is added
if (process.env.MULTICHAIN) {
  return;
}
```
## **Manual testing steps**

1. Run the new Eth Sign tests from `test/e2e/accounts/snap-account-signatures.spec.ts`
2. Make sure all the old tests still work properly

## **Related issues**

_**Q:**_ This has a lot of changed files, can you split it into different PRs?
_**A:**_ My work is already split into these other 3 related PRs: #21479 #21481 #20578.  All 4 PRs are designed to work together.  Splitting it up any more would be too laborious to keep updated.

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
